New Class Yoast_Form_Fieldset

afercia opened the pull request:
A very first pass, in order to have something to see and discuss. Some things still to do (including unit tests) and check. Any thoughts more than welcome, please consider this is not exactly my area of expertise :)

Current output in a quick test:

<img width="885" alt="screen shot 2016-05-18 at 18 38 50" src="https://cloud.githubusercontent.com/assets/1682452/15369609/acbc8196-1d33-11e6-8776-22c2989a7e79.png">

Fixes #4593.

afercia commented:
Should be refactored (maybe as method of `Yoast_Form`) and to use a separate method for the legend element, see #4680

afercia commented:
Was reviewing this today. So, this should be refactored but the problem here is how to build a method to have a fieldset to group together related form elements and at the same time don't refactor all the other `Yoast_Form` methods.

Consider for example something like this:
```
$yform->checkbox( 'importheadspace', __( 'Import from HeadSpace2?', 'wordpress-seo' ) );
$yform->checkbox( 'importaioseo', __( 'Import from All-in-One SEO?', 'wordpress-seo' ) );
```

these methods output some HTML and don't return anything so what should we pass to a new `Yoast_Form->fieldset()` method? We could try to catch the output buffer but it's something that should be avoided or maybe refactor `Yoast_Form` in order to introduce for each form element a new method that returns the HTML? e.g.  `get_checkbox()` and then `checkbox()` that just echoes `get_checkbox()`? Any feedback more than welcome.

moorscode commented on admin/views/form/fieldset.php:
```diff
@@ -0,0 +1,23 @@
+ * @var string $content             The fieldset content, i.e. a set of logically grouped form controls.
+ */
+?>
+<fieldset id="<?php esc_attr_e( $id ); ?>" <?php echo $fieldset_attributes; ?>>
```
`esc_attr_e` will translate the input. Better echo the `esc_attr` here.

moorscode commented on admin/views/form/fieldset.php:
```diff
@@ -0,0 +1,23 @@
+ */
+?>
+<fieldset id="<?php esc_attr_e( $id ); ?>" <?php echo $fieldset_attributes; ?>>
+	<legend id="<?php esc_attr_e( $id ); ?>-legend" <?php echo $legend_attributes; ?>><?php esc_html_e( $legend_content ); ?></legend>
```
See above ^

moorscode commented on admin/views/class-yoast-form-fieldset.php:
```diff
@@ -0,0 +1,136 @@
+		if ( ! empty( $attributes ) ) {
+			array_walk( $attributes, array( $this, 'parse_attribute' ) );
+
+			return implode( ' ', $attributes ) . ' ';
```
Why the extra space afterwards?

moorscode commented:
Looking at the flow, the content that will be inside the fieldset will have to be fetched first and fed into the fieldset class.

It feels like it would be a bit more logical to have a `startFieldset` and `endFieldset` kind of approach where everything in between will be encapsulated in the fieldset.

Internally we can just do an `ob_start` in the `startFieldset` write the header HTML in the buffer and close everything up on `endFieldset`. This would even work for nested fieldset.. if we would ever wanted that.

This would reduce the need for `get_*` functions, which are a good idea to have, but will have some considerations about current uses of the form class around the plugin and potentially being hooked into by other plugins.

afercia commented:
@moorscode thanks, was considering to completely refactor this. Maybe there's no need of a separate class (which was basically copied from `Yoast_Input_Select`) and `select.php` I think the  `esc_attr_e()` thing comes from there. Now that there's a `legend()` method maybe this is simple enough to be just a `Yoast_Form` method.

> it would be a bit more logical to have a startFieldset and endFieldset kind of approach
yep, tried that using output buffering but didn't look so nice 😞

atimmer commented:
This PR is the way forward. For now passing a string as the content for the fieldset is pragmatic, but it makes the API a little bit ugly when implementing it in an actual view. For example you'd have to do this to make it work:

```php
ob_start();
$yform->checkbox( 'importheadspace', __( 'Import from HeadSpace2?', 'wordpress-seo' ) );		  	$yform->checkbox( 'importheadspace', __( 'Import from HeadSpace2?', 'wordpress-seo' ) );
$yform->checkbox( 'importaioseo', __( 'Import from All-in-One SEO?', 'wordpress-seo' ) );		  	$yform->checkbox( 'importaioseo', __( 'Import from All-in-One SEO?', 'wordpress-seo' ) );
$yform->checkbox( 'importwoo', __( 'Import from WooThemes SEO framework?', 'wordpress-seo' ) );		  	$yform->checkbox( 'importwoo', __( 'Import from WooThemes SEO framework?', 'wordpress-seo' ) );
$yform->checkbox( 'importwpseo', __( 'Import from wpSEO', 'wordpress-seo' ) );		  	$yform->checkbox( 'importwpseo', __( 'Import from wpSEO', 'wordpress-seo' ) );
$fieldset_content = ob_get_clean();

$fieldset = new Yoast_Form_Fieldset( 'id', 'I am legend', $fieldset_content );
echo $fieldset->get_html();
```

Which is not really nice in a view, but we can do the same trick with start and end as in #5269 with an object:
```php
$fieldset = new Yoast_Form_Fieldset( 'id', 'I am legend', '' );
$fieldset->start();
$yform->checkbox( 'importheadspace', __( 'Import from HeadSpace2?', 'wordpress-seo' ) );		  	$yform->checkbox( 'importheadspace', __( 'Import from HeadSpace2?', 'wordpress-seo' ) );
$yform->checkbox( 'importaioseo', __( 'Import from All-in-One SEO?', 'wordpress-seo' ) );		  	$yform->checkbox( 'importaioseo', __( 'Import from All-in-One SEO?', 'wordpress-seo' ) );
$yform->checkbox( 'importwoo', __( 'Import from WooThemes SEO framework?', 'wordpress-seo' ) );		  	$yform->checkbox( 'importwoo', __( 'Import from WooThemes SEO framework?', 'wordpress-seo' ) );
$yform->checkbox( 'importwpseo', __( 'Import from wpSEO', 'wordpress-seo' ) );		  	$yform->checkbox( 'importwpseo', __( 'Import from wpSEO', 'wordpress-seo' ) );
$fieldset->end();
echo $fieldset->get_html();
```

For context, something like this is where we should be heading towards:

```php
$fieldset = new Yoast_Form_Fieldset( 'id', 'I am legend', array(
   new Yoast_Form_Choice( (...) ),
   new Yoast_Form_Choice( (...) ),
   new Yoast_Form_Choice( (...) ),
   new Yoast_Form_Choice( (...) ),
   new Yoast_Form_Choice( (...) ),
) );

echo $fieldset->get_html();
```

The reason for this is because when we have this structure we can not only create the HTML for the forms using the object, but we can just as easily generate JSON. The constructor of `Yoast_Form_Fieldset` would then look like this:

```php
class Yoast_Form_Fieldset implements `Yoast_Form_Field`

/**
 * @param string                       $fieldset_id          ID for the fieldset.
 * @param string                       $legend_content The legend text.
 * @param Yoast_Form_Field[] $fields                 The fields inside this fieldset.
 */
public function __construct( $fieldset_id, $legend_content, $fields ) {
```

So, what are the next steps?
1. Start with introducing a `Yoast_Form_Field` interface that is implemented by this new `Yoast_Form_Fieldset` class.
2. For now make it have a `start` and an `end` method to us output buffering to create the fieldset content.
3. When we have all field types as nice objects we can add a third argument for the fields inside a fieldset.

@afercia Can you do 1. and 2. in this PR?

afercia commented:
> Can you do 1. and 2. in this PR?
@atimmer I can try 😃

afercia commented:
First pass with the new approach and interface implementation. 

Have a small doubt about `esc_html( $legend_content )`: maybe when wrapping other form elements they're supposed to be already escaped?

Currently, this class can be used 1) with output buffering to wrap other form elements 2) passing a string as the content for the fieldset. Usage examples could be something like these:

```
$fieldset = new Yoast_Form_Fieldset( 'id', __( 'I am the legend' ), '' );
$fieldset->fieldset_add_attribute( 'class', 'blabla' );
$fieldset->fieldset_add_attribute( 'data-whatever', 'whatever' );
$fieldset->legend_add_attribute( 'class', 'blabla' );
$fieldset->legend_add_attribute( 'data-whatever', 'whatever' );
$fieldset->start();

$yform->checkbox( 'importheadspace', __( 'Import from HeadSpace2?', 'wordpress-seo' ) );
$yform->checkbox( 'importaioseo', __( 'Import from All-in-One SEO?', 'wordpress-seo' ) );
$yform->checkbox( 'importwoo', __( 'Import from WooThemes SEO framework?', 'wordpress-seo' ) );
$yform->checkbox( 'importwpseo', __( 'Import from wpSEO', 'wordpress-seo' ) );

$fieldset->end();
$fieldset->html();
```

or

```
$fieldset = new Yoast_Form_Fieldset( 'id', __( 'I am the legend' ), 'I am the content of the fieldset' );
$fieldset->fieldset_add_attribute( 'class', 'blabla' );
$fieldset->fieldset_add_attribute( 'data-whatever', 'whatever' );
$fieldset->legend_add_attribute( 'class', 'blabla' );
$fieldset->legend_add_attribute( 'data-whatever', 'whatever' );
$fieldset->html();
```

afercia commented:
Should we add some default attributes? e.g. default CSS classes.

atimmer commented on admin/views/class-yoast-form-fieldset.php:
```diff
@@ -0,0 +1,154 @@
+	 * @param string $attribute The name of the attribute to add.
+	 * @param string $value     The value of the attribute.
+	 */
+	public function fieldset_add_attribute( $attribute, $value ) {
```
You should call this `add_attribute`, because the object you are talking to is a fieldset object.

atimmer commented on admin/views/class-yoast-form-fieldset.php:
```diff
@@ -0,0 +1,154 @@
+	/**
+	 * @var array The fieldset HTML attributes.
+	 */
+	private $fieldset_attributes = array();
```
Given my comment on the `fieldset_add_attribute` function, you should call this `attributes`.

atimmer commented on admin/views/class-yoast-form-fieldset.php:
```diff
@@ -0,0 +1,154 @@
+	 *
+	 * @return string A space separated list of HTML attributes or empty string.
+	 */
+	private function get_attributes( $attributes ) {
```
`render_attributes` makes more sense and reserves `get_attributes` if we want to introduce a getter for the `attributes` property.

afercia replied:
Yep, I'd propose `get_attributes_html()` and also change the docblock here and for `parse_attribute()` as the summaries don't describe exactly what the methods do.

afercia commented:
Refactored a few things as suggested by @atimmer using output buffering twice until we have proper templating. 

- Fix usage of `render_view()` using output buffering twice
- Better methods and properties names
- Refactored methods to add attributes, they now accept an array of attributes and use `wp_parse_args()`
- Added default CSS class attributes to the fieldset and legend to use for default styling
- Added a `legend_hide()` utility method to hide the legend in an accessible way

moorscode commented on admin/views/form/fieldset.php:
```diff
@@ -0,0 +1,24 @@
+
+<fieldset id="<?php echo esc_attr( $id ); ?>"<?php echo $attributes; ?>>
+	<legend id="<?php echo esc_attr( $id ); ?>-legend"<?php echo $legend_attributes; ?>><?php echo esc_html( $legend_content ); ?></legend>
+	<?php echo $content . "\n"; ?>
```
Is that new-line really needed here? There is one after the `?>` as well.

moorscode commented:
CR Done 👍
It's all looking good, minor comment on the new-line, but anybody could remove it when checking out the branch for acceptance testing.

boblinthorst commented:
Acceptance: ok 👍
